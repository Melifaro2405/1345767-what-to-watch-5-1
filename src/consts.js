const ALL_GENRES = `All genres`;
const COUNT_SHOWN_GENRES = 10;
const COUNT_SHOWN_FILMS = 8;
const COUNT_LIKE_GENRE_FILMS = 4;

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const CommentLength = {
  MIN: 50,
  MAX: 400
};

const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  FILMS: `/films`,
  FILM_BY_ID: `/films/:id`,
  ADD_REVIEW: `/films/:id/review`,
  PLAYER: `/player`,
  PLAYER_BY_ID: `/player/:id`,
};

const APIRoute = {
  FILMS: `/films`,
  FAVORITE_FILMS: `/favorite`,
  PROMO_FILM: `/films/promo`,
  COMMENTS: `/comments`,
  LOGIN: `/login`
};

export {
  ALL_GENRES,
  COUNT_SHOWN_GENRES,
  COUNT_SHOWN_FILMS,
  COUNT_LIKE_GENRE_FILMS,
  AuthorizationStatus,
  CommentLength,
  AppRoute,
  APIRoute
};
