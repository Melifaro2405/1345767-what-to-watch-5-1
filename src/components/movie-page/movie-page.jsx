import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {filmProptypes} from "../../props-validation";
import Tabs from "../tabs/tabs";
import MoviesList from "../movies-list/movies-list";
import withActiveTab from "../../hocs/with-active-tab/with-active-tab";
import {connect} from "react-redux";
import {AppRoute, AuthorizationStatus, COUNT_LIKE_GENRE_FILMS} from "../../consts";
import Footer from "../footer/footer";
import UserBlock from "../user-block/user-block";
import ButtonChangeFilmStatus from "../button-change-film-status/button-change-film-status";
import {fetchFilmByID, updateFilmStatus} from "../../serviсes/api-actions";

const TabsWrapped = withActiveTab(Tabs);

class MoviePage extends PureComponent {
  constructor(props) {
    super(props);

    this._handleChangeFilmStatus = this._handleChangeFilmStatus.bind(this);
  }

  componentDidMount() {
    this._getUpdatedFilm();
  }

  componentDidUpdate(prevProps) {
    if (this.props.id !== prevProps.id) {
      this._getUpdatedFilm();
    }
  }

  _getUpdatedFilm() {
    const {id, getFilm, updateFilmByID} = this.props;

    getFilm(id)
    .then((film) => {
      updateFilmByID(film);
    });
  }

  _handleChangeFilmStatus() {
    const {id, film, changeFilmStatus, updateFilmByStatus} = this.props;
    const status = Number(!film.moreInfo.isAddToMyList);

    changeFilmStatus(id, status)
    .then((receivedFilm) => {
      updateFilmByStatus(receivedFilm);
    });
  }

  render() {
    const {film} = this.props;

    if (!film) {
      return null;
    }

    const {id, films, authorizationStatus} = this.props;

    const {moreInfo, preview} = film;
    const {backgroundColor, backgroundSrc, genre, releaseDate, posterSrc} = moreInfo;
    const isAddToMyList = moreInfo.isAddToMyList;
    const {title} = preview;
    const likeGenreFilms = films
    .filter((likeFilm) => genre === likeFilm.moreInfo.genre && likeFilm.id !== id)
    .slice(0, COUNT_LIKE_GENRE_FILMS);

    return (
      <React.Fragment>
        <section className="movie-card movie-card--full" style={{backgroundColor}}>
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img
                src={backgroundSrc}
                alt={title}
              />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <div className="logo">
                <Link to={AppRoute.ROOT} className="logo__link">
                  <span className="logo__letter logo__letter--1">W</span>
                  <span className="logo__letter logo__letter--2">T</span>
                  <span className="logo__letter logo__letter--3">W</span>
                </Link>
              </div>

              <UserBlock />
            </header>

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genre}</span>
                  <span className="movie-card__year">{releaseDate}</span>
                </p>

                <div className="movie-card__buttons">
                  <Link to={`${AppRoute.PLAYER}/${id}`} className="btn btn--play movie-card__button" type="button">
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s"/>
                    </svg>
                    <span>Play</span>
                  </Link>

                  <ButtonChangeFilmStatus isAddToMyList={isAddToMyList} onChangeFilmStatus={this._handleChangeFilmStatus}/>

                  {(authorizationStatus === AuthorizationStatus.AUTH) ?
                    <Link to={`${AppRoute.FILMS}/${id}/review`} className="btn movie-card__button">
                      Add review
                    </Link> : ``}

                </div>
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img
                  src={posterSrc}
                  alt={title}
                  width="218"
                  height="327"
                />
              </div>

              <div className="movie-card__desc">
                <TabsWrapped film={film} id={id} />
              </div>
            </div>
          </div>
        </section>
        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <MoviesList films={likeGenreFilms} />
          </section>

          <Footer />
        </div>
      </React.Fragment>
    );
  }
}

MoviePage.propTypes = {
  film: PropTypes.shape(filmProptypes),
  films: PropTypes.arrayOf(PropTypes.shape(filmProptypes)).isRequired,
  id: PropTypes.number.isRequired,
  getFilm: PropTypes.func.isRequired,
  updateFilmByID: PropTypes.func.isRequired,
  changeFilmStatus: PropTypes.func.isRequired,
  updateFilmByStatus: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired
};

const mapStateToProps = ({USER}) => ({
  authorizationStatus: USER.authorizationStatus
});

const mapDispatchToProps = (dispatch) => ({
  getFilm(id) {
    return dispatch(fetchFilmByID(id));
  },
  changeFilmStatus(id, status) {
    return dispatch(updateFilmStatus(id, status));
  }
});


export {MoviePage};
export default connect(mapStateToProps, mapDispatchToProps)(MoviePage);
