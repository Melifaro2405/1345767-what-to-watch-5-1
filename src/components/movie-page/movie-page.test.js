import React from 'react';
import {BrowserRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import {AuthorizationStatus} from '../../consts';
import {MoviePage} from './movie-page';

jest.mock(`../user-block/user-block`, () => `UserBlock`);
jest.mock(`../button-change-film-status/button-change-film-status`, () => `ButtonChangeFilmStatus`);
jest.mock(`../tabs/tabs`, () => `Tabs`);
jest.mock(`../movies-list/movies-list`, () => `MoviesList`);
jest.mock(`../footer/footer`, () => `Footer`);

const noop = () => {};
const noopPromise = () => {
  return {
    then: noop
  };
};

export const film = {
  id: 1,
  preview: {
    src: `img/the-grand-budapest-hotel-poster.jpg`,
    title: `The Grand Budapest Hotel`
  },
  moreInfo: {
    backgroundSrc: `img/the-grand-budapest-hotel-bg.jpg`,
    backgroundColor: `#ffffff`,
    posterSrc: `img/the-grand-budapest-hotel-poster.jpg`,
    genre: `Comedy`,
    releaseDate: 2014,
    playPreviewSrc: `https://some-link`,
    playVideoSrc: `https://some-link`,
    isAddToMyList: false
  },
  overview: {
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
      presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy,
      becomes Gustave's friend and protege.`,
    rating: 8.9,
    ratingCount: 240,
    director: `Wes Andreson`,
    actorsList: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`]
  },
  details: {
    runtime: 99
  }
};

export const films = [film];

describe(`Should MoviePage render correctly`, () => {
  test.each([
    [`with`, AuthorizationStatus.AUTH],
    [`without`, AuthorizationStatus.NO_AUTH],
  ])(`%s authorization`, (_expected, authorizationStatus) => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <MoviePage
              id={1}
              film={film}
              films={films}
              getFilm={noopPromise}
              updateFilmByID={noop}
              changeFilmStatus={noop}
              updateFilmByStatus={noop}
              authorizationStatus={authorizationStatus}
            />
          </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
