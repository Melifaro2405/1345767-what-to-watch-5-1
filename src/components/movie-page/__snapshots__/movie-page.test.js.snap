// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should MoviePage render correctly with authorization 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
          src="img/the-grand-budapest-hotel-bg.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <UserBlock />
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <ButtonChangeFilmStatus
              isAddToMyList={false}
              onChangeFilmStatus={[Function]}
            />
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <Tabs
            activeTab="Overview"
            film={
              Object {
                "details": Object {
                  "runtime": 99,
                },
                "id": 1,
                "moreInfo": Object {
                  "backgroundColor": "#ffffff",
                  "backgroundSrc": "img/the-grand-budapest-hotel-bg.jpg",
                  "genre": "Comedy",
                  "isAddToMyList": false,
                  "playPreviewSrc": "https://some-link",
                  "playVideoSrc": "https://some-link",
                  "posterSrc": "img/the-grand-budapest-hotel-poster.jpg",
                  "releaseDate": 2014,
                },
                "overview": Object {
                  "actorsList": Array [
                    "Bill Murray",
                    "Edward Norton",
                    "Jude Law",
                    "Willem Dafoe",
                    "Saoirse Ronan",
                  ],
                  "description": "In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
      presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy,
      becomes Gustave's friend and protege.",
                  "director": "Wes Andreson",
                  "rating": 8.9,
                  "ratingCount": 240,
                },
                "preview": Object {
                  "src": "img/the-grand-budapest-hotel-poster.jpg",
                  "title": "The Grand Budapest Hotel",
                },
              }
            }
            id={1}
            onClickTab={[Function]}
          />
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <MoviesList
        films={Array []}
      />
    </section>
    <Footer />
  </div>,
]
`;

exports[`Should MoviePage render correctly without authorization 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
          src="img/the-grand-budapest-hotel-bg.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <UserBlock />
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <ButtonChangeFilmStatus
              isAddToMyList={false}
              onChangeFilmStatus={[Function]}
            />
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <Tabs
            activeTab="Overview"
            film={
              Object {
                "details": Object {
                  "runtime": 99,
                },
                "id": 1,
                "moreInfo": Object {
                  "backgroundColor": "#ffffff",
                  "backgroundSrc": "img/the-grand-budapest-hotel-bg.jpg",
                  "genre": "Comedy",
                  "isAddToMyList": false,
                  "playPreviewSrc": "https://some-link",
                  "playVideoSrc": "https://some-link",
                  "posterSrc": "img/the-grand-budapest-hotel-poster.jpg",
                  "releaseDate": 2014,
                },
                "overview": Object {
                  "actorsList": Array [
                    "Bill Murray",
                    "Edward Norton",
                    "Jude Law",
                    "Willem Dafoe",
                    "Saoirse Ronan",
                  ],
                  "description": "In the 1930s, the Grand Budapest Hotel is a popular European ski resort,
      presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy,
      becomes Gustave's friend and protege.",
                  "director": "Wes Andreson",
                  "rating": 8.9,
                  "ratingCount": 240,
                },
                "preview": Object {
                  "src": "img/the-grand-budapest-hotel-poster.jpg",
                  "title": "The Grand Budapest Hotel",
                },
              }
            }
            id={1}
            onClickTab={[Function]}
          />
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <MoviesList
        films={Array []}
      />
    </section>
    <Footer />
  </div>,
]
`;
